### PR TITLE
evdns: recreate ns socket after ENOTCONN

### LINKED ISSUE #265: evdns loops forever if UDP socket gets ENOTCONN
If `recvfrom` results in `ENOTCONN`, the socket remains readable which triggers another `recvfrom`. The observed behavior is 100% CPU use.

This occurs on iOS (kqueue) after the process has been backgrounded for a long time (~300 seconds) and then resumed. All sockets, TCP and UDP, seem to get `ENOTCONN` and must be closed.

### PR DESCRIPTION
**Pull request by okhowang**
with ghazel's [patch](https://github.com/libevent/libevent/issues/265#issuecomment-123449315).

Closes: #265

### CODE REVIEW

**Comment by okhowang** [comment]
there is no other change except `evdns.c`.
failed tests is likely to be not robust?

**Comment by azat** [comment]
>failed tests is likely to be not robust?

yes, you can safely ignore them

File: evdns.c [hunk 1]
```diff
@@ -731,9 +731,32 @@ nameserver_failed(struct nameserver *const ns, const char *msg) {
 	ns->state = 0;
 	ns->failed_times = 1;
 
-	disconnect_and_free_connection(ns->connection);
-	ns->connection = NULL;
+	if (ns->connection) {
+		disconnect_and_free_connection(ns->connection);
+		ns->connection = NULL;
+	} else {
+		// XXX: https://github.com/libevent/libevent/issues/265
```
Comment (azat) [inline evdns.c:738]: This should be done only for `ENOTCONN` not always

Comment (okhowang) [reply evdns.c:749]: fixed

**Comment by okhowang** [comment]
> Can you provide system calls trace for this problem? (with evdns logging enabled)

I'm sorry, I can't reproduce the error in my device.
but it appear in our users's device which can't enable evdns log.

File: evdns.c [hunk 2]
```diff
@@ -742,9 +742,32 @@ nameserver_failed(struct nameserver *const ns, const char *msg) {
...
+					base->global_outgoing_addrlen) < 0) {
+				log(EVDNS_LOG_WARN, "Couldn't bind to outgoing address");
+			}
+		}
 
+		event_del(&ns->event);
+		event_assign(&ns->event, ns->base->event_base, ns->socket,
+			EV_READ | (ns->write_waiting ? EV_WRITE : 0) | EV_PERSIST,
+			nameserver_ready_callback, ns);
+		if (!base->disable_when_inactive && event_add(&ns->event, NULL) < 0) {
```
Comment (azat) [inline evdns.c:768]: Please cleanup the code, empty if branch is not ok, let's add log message at least

**Comment by azat** [comment]
>I'm sorry, I can't reproduce the error in my device.

Ok, this doesn't look hurmful, so this can be merged
But this doesn't have a reproducer, I wish we could at it..

File: evdns.c [hunk 3]
```diff
@@ -742,9 +742,32 @@ nameserver_failed(struct nameserver *const ns, const char *msg) {
 	ns->state = 0;
 	ns->failed_times = 1;
 
-	disconnect_and_free_connection(ns->connection);
-	ns->connection = NULL;
+	if (ns->connection) {
+		disconnect_and_free_connection(ns->connection);
+		ns->connection = NULL;
+	} else if (err == ENOTCONN) {
+		// XXX: https://github.com/libevent/libevent/issues/265
```
Comment (azat) [inline evdns.c:749]: And please can you describe the problem in the comments, providing link is not enough

File: evdns.c (hunk 2 above)
Comment (okhowang) [reply evdns.c:768]: fixed

File: evdns.c (hunk 3 above)
Comment (okhowang) [reply evdns.c]: fixed

File: evdns.c [hunk 4]
```diff
@@ -742,9 +742,37 @@ nameserver_failed(struct nameserver *const ns, const char *msg) {
...
-	disconnect_and_free_connection(ns->connection);
-	ns->connection = NULL;
+	if (ns->connection) {
+		disconnect_and_free_connection(ns->connection);
+		ns->connection = NULL;
+	} else if (err == ENOTCONN) {
+		/* XXX: on iOS, when app become background, socket may become unusable.
+		 * But when it will be also readable for kqueue/poll.
+		 * It lead infinite loop of recvfrom -> error -> recvfrom -> error.
+		 * we recreate socket in the ENOTCONN error to break this infinite loop.
```
Comment (azat) [inline evdns.c:752]: Sorry, message is not clear enough for me, I'm suggesting you to just copy description from the linked issue

File: evdns.c [hunk 5]
```diff
@@ -742,9 +742,37 @@ nameserver_failed(struct nameserver *const ns, const char *msg) {
...
+				log(EVDNS_LOG_WARN, "Couldn't bind to outgoing address");
+			}
+		}
 
+		event_del(&ns->event);
+		event_assign(&ns->event, ns->base->event_base, ns->socket,
+			EV_READ | (ns->write_waiting ? EV_WRITE : 0) | EV_PERSIST,
+			nameserver_ready_callback, ns);
+		if (!base->disable_when_inactive && event_add(&ns->event, NULL) < 0) {
+			log(EVDNS_LOG_WARN, "Couldn't add event");
```
Comment (azat) [inline evdns.c:773]: This is not warm, and can you describe which event?

Comment (okhowang) [reply evdns.c:773]: event type added

File: evdns.c (hunk 4 above)
Comment (okhowang) [reply evdns.c:752]: copied

Comment (azat) [reply evdns.c:752]: Have you pushed the changes? If you want to, I can finish by myself (to reduce latency for interaction) or I can wait, it's up to you

Comment (okhowang) [reply evdns.c]: I'm sorry for forgeting push.
pushed currently

**Pull request merged**